pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,885 @@
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		if !apierrors.IsNotFound(err) {
+			c.queue.AddRateLimited(key)
+			klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+		} else {
+			c.queue.Forget(key)
+			delete(c.captures, key)
+		}
+	}
+	return true
+}
+
+func (c *Controller) cleanupPacketCapture(pcName string) {
+	path := nameToPath(pcName)
+	exist, err := afero.Exists(defaultFS, path)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if path exists", "path", path)
+	}
+	if !exist {
+		return
+	}
+	if err := defaultFS.Remove(path); err == nil {
+		klog.V(2).InfoS("Deleted the captured pcap file successfully", "name", pcName, "path", path)
+	} else {
+		klog.ErrorS(err, "Failed to delete the captured pcap file", "name", pcName, "path", path)
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, nil, fmt.Errorf("the packet file %s already exists. This may be caused by an unexpected termination", filePath)
+	} else if os.IsNotExist(err) {
+		file, err = defaultFS.Create(filePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+// the agent on this Node will not perform the capture.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+
+	go func() error {
+		var err error
+		pcState := &packetCaptureState{name: name}
+		defer func() {
+			if err != nil {
+				c.cleanupPacketCapture(name)
+				klog.ErrorS(err, "PackageCapture failed", "name", name)
+				err := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", err)
+				if err != nil {
+					klog.ErrorS(err, "failed to update PacketCapture status")
+				}
+			}
+		}()
+		pc, err := c.packetCaptureLister.Get(name)
+		if err != nil {
+			return err
+		}
+
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil
+		}
```

Comment:
the early return cases (e.g. because of errors) are not handled correctly:
- we never update the state (`c.numCaptures`, `c.captures[name]`) and never signal the condition variable
- we don't have a retry mechanism; I think that in case of an error, we should be able to retry, which means adding the capture back to the workqueue

Any error will cause the state to become invalid (with the current code). We need a test for the error case if possible.

Reply:
I have actually updated in the processPacketCaptureItem：

```
	if err := c.syncPacketCapture(key); err == nil {
		c.queue.Forget(key)
	} else {
		if !apierrors.IsNotFound(err) {
			c.queue.AddRateLimited(key)
			klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
		} else {
			c.queue.Forget(key)
			delete(c.captures, key)
		}
	}
```

only NotFound error will be ignored, others will requeue. I'm not sure if that's enough?